### PR TITLE
*: improve RUNTIME_CHECK and RUNTIME_ASSERT

### LINKED ISSUE #5444: Better `RUNTIME_CHECK` and `RUNTIME_ASSERT`
### `RUNTIME_CHECK`

Current `RUNTIME_CHECK` usage is:

```
RUNTIME_CHECK(a != b, ExceptionType, "{} != {}", a, b);
```

It's not clear for users to know which exception will be thrown. A better way is to explicit write code for generating an exception object like:

```
RUNTIME_CHECK(a != b, ExceptionType("{} != {}", a, b));
```

And the new style could enable better compilation checks.

### `RUNTIME_ASSERT`

Current `RUNTIME_ASSERT` always needs a `logger` and make it hard for replacing existent `assert`s. We should let it work without user provided logger (in which case a default logger is used).

### PR DESCRIPTION
**Pull request by fuzhe1989**
Signed-off-by: fuzhe1989 <fuzhe1989@gmail.com>

### What problem does this PR solve?

Issue Number: close #5444 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by fuzhe1989** [comment]
any update? @JaySon-Huang @SeaRise @gengliqi

File: dbms/src/Common/Exception.h [hunk 1]
```diff
@@ -172,28 +162,123 @@ std::enable_if_t<std::is_pointer_v<T>, T> exception_cast(std::exception_ptr e)
...
+{
+    return Poco::Message(
+        logger_name,
+        generateFormattedMessage(condition, std::forward<Args>(args)...),
+        Poco::Message::PRIO_FATAL,
+        filename,
+        lineno);
+}
+
+Poco::Logger * getDefaultFatalLogger();
```
Comment (gengliqi) [inline dbms/src/Common/Exception.h:187]: How about returning `LoggerPtr` then 
```cpp
inline void log(const char * filename, int lineno, const char * condition, Poco::Logger * logger)
```
and 
```cpp
inline void log(const char * filename, int lineno, const char * condition, Poco::Logger * logger, const char * fmt_str, Args &&... args)
```
can be removed.

Comment (fuzhe1989) [reply dbms/src/Common/Exception.h:187]: hmmm ... looks like we don't need to support `Poco::Loger *` at the first revision?

Comment (gengliqi) [reply dbms/src/Common/Exception.h:187]: I guess so...

Comment (fuzhe1989) [reply dbms/src/Common/Exception.h]: done

File: dbms/src/Common/Exception.h [hunk 2]
```diff
@@ -172,28 +162,98 @@ std::enable_if_t<std::is_pointer_v<T>, T> exception_cast(std::exception_ptr e)
...
+    }
+}
+
+inline void log(const char * filename, int lineno, const char * condition)
+{
+    log(filename, lineno, condition, getDefaultFatalLogger());
+}
+
+template <typename... Args>
+inline void log(const char * filename, int lineno, const char * condition, const LoggerPtr & logger, const char * fmt_str, Args &&... args)
```
Comment (gengliqi) [inline dbms/src/Common/Exception.h:204]: How about removing `const char * fmt_str`? 
If so, `inline void log(const char * filename, int lineno, const char * condition, const LoggerPtr & logger)` can be removed too.

File: dbms/src/Common/Exception.h [hunk 3]
```diff
@@ -172,28 +162,98 @@ std::enable_if_t<std::is_pointer_v<T>, T> exception_cast(std::exception_ptr e)
...
+            lineno,
+            condition,
+            fmt_str,
+            std::forward<Args>(args)...);
+        logger->log(message);
+    }
+}
+
+template <typename... Args>
+inline void log(const char * filename, int lineno, const char * condition, const char * fmt_str, Args &&... args)
```
Comment (gengliqi) [inline dbms/src/Common/Exception.h:220]: ditto.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

File: dbms/src/Common/Exception.h (hunk 2 above)
Comment (fuzhe1989) [reply dbms/src/Common/Exception.h]: done

File: dbms/src/Common/Exception.h (hunk 3 above)
Comment (fuzhe1989) [reply dbms/src/Common/Exception.h:210]: this `fmt_str` is used for overloadding from `const LoggerPtr &` or this version will always be chosen even when there's a `const LoggerPtr &`.

**Review by gengliqi — APPROVED** [review]

**Pull request merged**